design/restore-with-EnableAPIGroupVersions-feature.md

```diff
@@ -0,0 +1,193 @@
+  
+  Once the version to restore is chosen, the file path to the backed up resource in the tarball will be modified such that it points to the resources' chosen API group version. If no version is found in common between the source and target clusters, the chosen version will default to the source cluster's preferred version (the version being restored currently without the changes proposal here). Restore will be allowed to continue as before.
+
+## Detailed Design
+
+There are six objectives to achieve the above stated goals:
+
+1. Determine if the APIGroupVersionsFeatureFlag feature flag has been enabled.
+1. List the backed up API group versions.
+1. List the API group versions supported by the target cluster.
+1. Get the user-defined version priorities.
+1. Use a priority system to determine which version to restore. The source preferred version will be the default if the priorities fail.
+1. Modify the paths to the backup files in the tarball in the resource restore process.
+
+### Objective 1: Determine if the APIGroupVersionsFeatureFlag feature flag has been enabled
+
+For restore to be able to choose from multiple supported backed up versions, the feature flag must have been enabled during both the backup and restore processes. And the check to see if the feature flag is/was enabled must be done for both processes.
+
+The reason for checking for the feature flag during restore is to ensure the user would like to restore a version that is among potentially multiple versions backed up with the feature flag enabled.
+
+The reason for checking to see if the feature flag was enabled during backup is to ensure the changes made by this proposed design is backward compatible. Only with Velero version 1.4 and forward was Format Version 1.0.0 used to structure the backup directories. Format Version 1.0.0 is required for the restore process proposed in this design doc to work. Before v1.4, the backed up files were in a directory structure that will not be recognized by the proposed code changes. Therefore, restore should not attempt to restore from multiple versions as they will not exist.
+
+Checking if the feature flag is enabled during restore is straightforward using `features.IsEnabled(velerov1api.APIGroupVersionsFeatureFlag)`.
+
+Checking if the feature flag was enabled during backup can be done indirectly by checking the directory structure and version directory names. For backups generated using Format Version 1.0.0 (required for the changes proposed here), there will always be a version directory with `-preferredversion` in the directory name. Only one resource will need to be checked to determine if the feature flag was enabled during backup.
```

Comment:
I think this paragraph is a little inaccurate.

A backup can be format 1.1.0 without the feature flag enabled, and should have the `-preferredversion` directory even without the feature flag. 

At backup time currently, the feature flag controls whether all versions are downloaded. By default, Velero only downloads the preferred version and stores it twice; once under the `-preferredversion` directory, once under a version-less directory.

Therefore, I think this could be simplified to only operate on `Backup` objects with `Status.FormatVersion == "1.1.0"`.  [Here's the actual const value](https://github.com/vmware-tanzu/velero/blob/main/pkg/backup/backup.go#L58).

Reply:
I had submitted a PR #2901 previously to fix the describe backup format version incorrectly (which was released at v1.5.0).
So, the only concern for me is if the source cluster uses Velero 1.4.0 to performs backup, then we can't use the format version to do the check.

Reply:
Thank you for the suggestion and another very good catch.

You're right that the paragraph is inaccurate. I will change it to say that we'll check for `Status.FormatVersion == "1.1.0"`. I realize from your observation that it is ok that during backup, the feature flag was not enabled. What's important are these two things:
(1) Feature flag is enabled during restore
(2) FormatVersion == 1.1.0

I'll be sure to change my wording to reflect those two conditions.

Reply:
@jenting From conversations with Rafael, Velero won't be able to do this migration with backups made prior to v1.4.0 anyway. The feature flag and support for any API version besides the preferred version was only introduced in v1.4.0.

Also, see [line 21 of this document](https://github.com/vmware-tanzu/velero/pull/3050/files#diff-85b71029d5725d4b9e77b52b2276f38caa0086332e139cbe9f49b9b7abb7401cR21), which states that doing migration with Velero versions 1.3.x or older is a non-goal.

@codegold79 Thanks! The important thing about the feature flag during backup is whether or not more than 1 version is backed up, so that version negotiation can take place. Having a backup of Format 1.1.0 means you'll get the `-preferredversion` directory, but it doesn't guarantee you'll get multiple API versions - that only happens when the feature flag is on. So I think that may be useful to check for still, though it's _also_ possible that a given API only has 1 version, anyway...which makes that heuristic kind of unreliable.

Off the top of my head, I don't think we write the feature flags that were enabled anywhere in the backup data, but maybe that would be useful so we don't have to use these kinds of nondeterministic heuristics.

Reply:
@jenting Sorry, I misunderstood your comment.

The code wouldn't have to use `velero describe`, it would use the `FormatVersion` field directly, which should still be reliable. The PR you referenced is only about the display on `velero describe`, not whether the `FormatVersion` field gets set correctly.

Reply:
Thanks for the detailed explanation 🙏, I have no other concerns.